api/src/main/java/org/apache/iceberg/BatchScanAdapter.java

```diff
@@ -0,0 +1,142 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.Collection;
+import java.util.concurrent.ExecutorService;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.io.CloseableIterable;
+
+class BatchScanAdapter implements BatchScan {
+
+  private final TableScan tableScan;
+
+  BatchScanAdapter(TableScan tableScan) {
+    this.tableScan = tableScan;
+  }
+
+  @Override
+  public Table table() {
+    return tableScan.table();
+  }
+
+  @Override
+  public BatchScan useSnapshot(long snapshotId) {
+    return new BatchScanAdapter(tableScan.useSnapshot(snapshotId));
+  }
+
+  @Override
+  public BatchScan useRef(String ref) {
+    return new BatchScanAdapter(tableScan.useRef(ref));
+  }
+
+  @Override
+  public BatchScan asOfTime(long timestampMillis) {
+    return new BatchScanAdapter(tableScan.asOfTime(timestampMillis));
+  }
+
+  @Override
+  public Snapshot snapshot() {
+    return tableScan.snapshot();
+  }
+
+  @Override
+  public BatchScan option(String property, String value) {
+    return new BatchScanAdapter(tableScan.option(property, value));
+  }
+
+  @Override
+  public BatchScan project(Schema schema) {
+    return new BatchScanAdapter(tableScan.project(schema));
+  }
+
+  @Override
+  public BatchScan caseSensitive(boolean caseSensitive) {
+    return new BatchScanAdapter(tableScan.caseSensitive(caseSensitive));
+  }
+
+  @Override
+  public boolean isCaseSensitive() {
+    return tableScan.isCaseSensitive();
+  }
+
+  @Override
+  public BatchScan includeColumnStats() {
+    return new BatchScanAdapter(tableScan.includeColumnStats());
+  }
+
+  @Override
+  public BatchScan select(Collection<String> columns) {
+    return new BatchScanAdapter(tableScan.select(columns));
+  }
+
+  @Override
+  public BatchScan filter(Expression expr) {
+    return new BatchScanAdapter(tableScan.filter(expr));
+  }
+
+  @Override
+  public Expression filter() {
+    return tableScan.filter();
+  }
+
+  @Override
+  public BatchScan ignoreResiduals() {
+    return new BatchScanAdapter(tableScan.ignoreResiduals());
+  }
+
+  @Override
+  public BatchScan planWith(ExecutorService executorService) {
+    return new BatchScanAdapter(tableScan.planWith(executorService));
+  }
+
+  @Override
+  public Schema schema() {
+    return tableScan.schema();
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public CloseableIterable<ScanTask> planFiles() {
+    CloseableIterable<?> fileScanTasks = tableScan.planFiles();
```

Comment:
It is a trick that uses type erasure. As `FileScanTask` extends `ScanTask` and `CloseableIterable` is immutable, this seems safe.

Reply:
I think it would be better to use `? extends ScanTask` as the type rather than just `?`. That's minor, but at least it is obvious to the reader what is happening here and that it is safe, rather than needing to look at `planFiles` and then then check that `FileScanTask` extends `ScanTask`.

Reply:
Let try it out locally. I'll add a comment at least.

Reply:
Done.